octree/basic.go

```diff
@@ -0,0 +1,143 @@
+package octree
+
+import (
+	"context"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+
+	pc "go.viam.com/rdk/pointcloud"
+)
+
+// basicOctree is a data structure that represents a basic octree structure with information regarding center
+// point, side length and node data.
+type basicOctree struct {
+	logger golog.Logger
+	node   basicOctreeNode
+	center r3.Vector
+	side   float64
+	size   int32
+	meta   pc.MetaData
+}
+
+// basicOctreeNode is a struct comprised of the type of node, children nodes (should they exist) and the pointcloud's
+// PointAndData datatype representing a point in space.
+type basicOctreeNode struct {
+	nodeType NodeType
+	tree     []*basicOctree
+	point    pc.PointAndData
+}
+
+// New creates a new basic octree with specified center, side and metadata.
+func New(ctx context.Context, center r3.Vector, sideLength float64, logger golog.Logger) (Octree, error) {
+	if sideLength <= 0 {
+		return nil, errors.Errorf("invalid side length (%.2f) for octree", sideLength)
+	}
+
+	octree := &basicOctree{
+		logger: logger,
+		node:   newLeafNodeEmpty(),
+		center: center,
+		side:   sideLength,
+		size:   0,
+	}
+	octree.meta = pc.NewMetaData()
+	return octree, nil
+}
+
+// Size returns the number of points stored in the octree's metadata.
+func (octree *basicOctree) Size() int {
+	return int(octree.size)
+}
+
+// Set checks if the point to be added is a valid point for a basic octree to contain based on its center and side
+// length. It then recursively iterates through the tree until it finds the appropriate node to add it too. If the
+// found node contains a point already, it will split the node into octants and will add both the old point and new
+// one to the newly created children trees.
+func (octree *basicOctree) Set(p r3.Vector, d pc.Data) error {
+	if (pc.PointAndData{P: p, D: d} == pc.PointAndData{}) {
+		octree.logger.Debug("no data given, skipping insertion")
+		return nil
+	}
+
+	if !checkPointPlacement(octree.center, octree.side, p) {
+		return errors.New("error point is outside the bounds of this octree")
+	}
```

Comment:
This introduces a o(n) potential complexity to any point addition as it may need to resize and rearrange all points in the tree. Current pointclouds are designed to be treated as immutable objects (they aren't but it is something Bijan has been working on), any changes to a pointcloud are done by creating a new pointcloud. I believe we want to take a similar approach with octrees. 

Additionally since most current usages for octrees currently have them being generated from a pointcloud/serialized octree there will be no need for adding additional points that may be outside both the bounds of the octree and of the pointcloud

Reply:
Additonal information: One of the major reasons pointclouds are treated as immutable is that there is no way to delete points. The Unset() function does not exist for pointclouds nor for octrees. This will be particularly important for SLAM as if we choose to update certain sections of an octree, data may be removed or probability values changed, without a way to actually remove points.

My recommendation is that we leave it as an immutable object to parallel the current implementation in pointcloud. If we want to make both objects mutable then this should happen in pointcloud first to add such functions as `Unset` and `Join` to the interface, before being tackled by octree. If we wish to make octrees mutable while keeping pointclouds immutable, then a new scope will have to be made to account for this and add the necessary functions.

Associated JIRA ticket: https://viam.atlassian.net/jira/software/c/projects/DATA/boards/28?modal=detail&selectedIssue=DATA-310